alphafold3_pytorch/alphafold3.py

```diff
@@ -3361,17 +3362,31 @@ def forward(
         return_pae_logits = exists(pae_labels)
 
         if calc_diffusion_loss and should_call_confidence_head:
+            
+            # rollout
+            pred_atom_pos = self.edm.sample(
+                num_sample_steps = num_rollout_steps,
+                atom_feats = atom_feats,
+                atompair_feats = atompair_feats,
+                atom_mask = atom_mask,
+                mask = mask,
+                single_trunk_repr = single,
+                single_inputs_repr = single_inputs,
+                pairwise_trunk = pairwise,
+                pairwise_rel_pos_feats = relative_position_encoding,
+                residue_atom_lens = residue_atom_lens
+            )
 
             if self.packed_atom_repr:
                 pred_atom_pos = einx.get_at('b [m] c, b n -> b n c', denoised_atom_pos, residue_atom_indices)
```

Comment:
@gitabtion hey Heng, i don't think you accounted for `pred_atom_pos` being overwritten here

Reply:
As mentioned in section 4.1, the `x_pred` used by the Auxiliary heads consists of coordinates generated through mini-rollouts, which aligns with our understanding. The denoised coordinates produced during the diffusion training steps are intermediate states in the generation process and differ significantly from the final denoised coordinates generated during inference. On the other hand, the Confidence heads are designed to evaluate the confidence of the generated structures. Therefore, during the training of the Confidence heads, it is appropriate to use coordinates generated from complete sampling (rollout) as input features and compute labels such as pLDDT, PAE, PDE, etc., using these coordinates and the ground truth (which is yet to be implemented).

The utilization of rollout-generated coordinates to replace `pred_atom_pos` serves as input features for the Confidence heads. Additionally, due to the presence of symmetry in certain atoms or chains, we also need to implement a procedure to align and adjust the order of the ground truth's atoms or chains using coordinates generated through rollouts.

Reply:
@lucidrains ah, I made a mistake! It should be named `denoised_atom_pos` instead of `pred_atom_pos`.

Reply:
indeed, thanks!